packages/apps/plugins/plugin-graph/src/graph/graph.ts

```diff
@@ -0,0 +1,189 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { RevertDeepSignal, deepSignal } from 'deepsignal/react';
+import get from 'lodash.get';
+
+import { SendIntent } from '@braneframe/plugin-intent';
+import { EventSubscriptions } from '@dxos/async';
+
+import { Graph } from './types';
+
+export class SessionGraph implements Graph {
+  private readonly _nodeBuilders = new Set<Graph.NodeBuilder>();
+  private readonly _unsubscribe = new Map<string, EventSubscriptions>();
+  private readonly _root: Graph.Node = this._createNode({ id: 'root', label: 'Root' });
+  // TODO(wittjosiah): Should this support multiple paths to the same node?
+  private readonly _index = deepSignal<{ [key: string]: string[] }>({});
+
+  private _sendIntent?: SendIntent;
+
+  get root(): Graph.Node {
+    return this._root;
+  }
+
+  find(id: string): Graph.Node | undefined {
+    const path = this._index[id];
+    if (!path) {
+      return undefined;
+    }
+
+    return path.length > 0 ? get(this._root, path) : this._root;
+  }
+
+  traverse({ from = this._root, direction = 'down', predicate, onVisitNode }: Graph.TraverseOptions): void {
+    if (!predicate || predicate(from)) {
+      onVisitNode?.(from);
+    }
+
+    if (direction === 'down') {
+      Object.values(from.children).forEach((child) => this.traverse({ from: child, predicate, onVisitNode }));
+    } else if (direction === 'up' && from.parent) {
+      this.traverse({ from: from.parent, direction, predicate, onVisitNode });
+    }
+  }
+
+  registerNodeBuilder(builder: Graph.NodeBuilder): void {
```

Comment:
I might try to keep the builders outside of the graph so that the graph itself is "inert" and keep it as simple as possible (because it is likely going to get more and more complex).

pseudocode:

```ts
const builder = new GraphBuilder().addNodeBuilder().addNodeBuilder();
const graph = builder.build();
```

Reply:
So...I started with it being separate, but the builders and the graph are quite intertwined. I think it might be doable to pull them apart if not for the fact that we filter the builders as you build up the graph so nodes aren't passed to builders which contributed to their ancestors. Maybe it's still possible? I'm open to looking at it again, but definitely a difficult one.

Reply:
concur, let's not make a change like this until we understand clearly the imminent end user scenario that would require that change.